rir/src/optimization/cleanup.h

```diff
@@ -187,8 +187,15 @@ class BCCleanup : public InstructionDispatcher::Receiver {
                 bubbleUp = bubbleUp - 1;
                 auto cur = *bubbleUp;
                 // We cannot move the guard across those instructions
-                if (cur.is(Opcode::label) || !cur.isPure() || cur.isReturn())
-                    break;
+                if (cur.is(Opcode::label) || !cur.isPure() || cur.isReturn()) {
+                    if (!cur.is(Opcode::stvar_)) {
+                        break;
+                    }
+                    // stvar that modiefies the guarded value is a barrier,
+                    // but otherwise we can push through
+                    if (Pool::get(cur.immediate.pool) == expected)
```

Comment:
Doesn't this compare a symbol (`stvar_`'s immediate) with the actual guarded function (`guard_fun_`'s expected immediate)?

Reply:
ups, good catch.
unfortunately I can't write a testcase, since currently our guard just aborts the program :)